cr8/clients.py

```diff
@@ -216,18 +216,25 @@ def _verify_ssl_from_first(hosts):
 
 
 class AsyncpgClient:
-    def __init__(self, hosts, pool_size=25):
+    def __init__(self, hosts, pool_size=25, session_settings=None):
         self.dsn = _to_dsn(hosts)
         self.pool_size = pool_size
         self._pool = None
         self.is_cratedb = True
+        self.session_settings = session_settings or {}
 
     async def _get_pool(self):
+
+        async def set_session_settings(conn):
+            for setting, value in self.session_settings.items():
+                await conn.execute(f'set {setting}={value}')
+
```

Comment:
As mentioned in chat, this needs some kind of check to see if the connection is new or getting re-used.

E.g. running a spec like:

```toml
[session_settings]
application_name = 'my_app'
timezone = 'UTC'


[[queries]]
name = "test"
statement = "select count(*) from sys.summits"
concurrency = 10
iterations = 1000
```


with `asyncpg://...` will result in:


```
cr> select stmt, count(*) from sys.jobs_log group by stmt order by 2 desc ;
+-------------------------------------------------------------------------+----------+
| stmt                                                                    | count(*) |
+-------------------------------------------------------------------------+----------+
| set timezone=UTC                                                        |     1002 |
| set application_name=my_app                                             |     1002 |
| select count(*) from sys.summits                                        |     1000 |
| select version from sys.nodes                                           |        1 |
| select stmt, count(*) from sys.jobs_log group by stmt order by 2 desc ; |        1 |
| ANALYZE                                                                 |        1 |
+-------------------------------------------------------------------------+----------+
```

The `set` calls should at most run 10 times in this case

Reply:
I fixed it for `asyncpg` with `init`:

+--------------------------------------------------------------------------------------------------+----------+
| stmt                                                                                             | count(*) |
+--------------------------------------------------------------------------------------------------+----------+
| select count(*) from doc.t1 join doc.t2 on t1.id = t2.id                                         |      100 |
| set optimizer_equi_join_to_lookup_join=true                                                      |       26 |
| refresh table t1, t2                                                                             |        1 |
| optimize table t2 with (max_num_segments = 1)                                                    |        1 |
| ANALYZE                                                                                          |        1 |
| create table t1 (id int primary key,a text,b text)                                               |        1 |
| optimize table t1 with (max_num_segments = 1)                                                    |        1 |
| select stmt, count(*) from sys.jobs_log group by stmt order by 2 desc                            |        1 |
| select version from sys.nodes                                                                    |        1 |
| insert into t2(ts,id,c) select now(),b%7000,b%6300 from generate_series(1,1000) a(b)             |        1 |
| drop table if exists t1                                                                          |        1 |
| drop table if exists t2                                                                          |        1 |
| insert into t1 (id,a,b) select b,(b%300)::TEXT,(b%450)::TEXT from generate_series(1,100000) a(b) |        1 |
| create table t2 (ts timestamp,id int,c double)                                                   |        1 |
+--------------------------------------------------------------------------------------------------+----------+